src/shared/errors.ts

```diff
@@ -275,18 +275,102 @@ export function getTelemetryReason(error: unknown | undefined): string | undefin
     return 'Unknown'
 }
 
-export function isAwsError(error: unknown | undefined): error is AWSError {
+/**
+ * Determines the appropriate error message to display to the user.
+ *
+ * We do not want to display every error message to the user, this
+ * resolves what we actually want to show them based off the given
+ * input.
+ */
+export function resolveErrorMessageToDisplay(error: unknown, defaultMessage: string): string {
+    let message = error instanceof ToolkitError ? error.message : defaultMessage
+    // We want to explicitly show certain AWS Error messages if they are raised
+    message = findPrioritizedAwsError(error)?.message ?? message
+    return message
```

Comment:
Just to check, before this whole flow starts we log the entire ToolkitError, showing the entire context: https://github.com/aws/aws-toolkit-vscode/blob/d6de7add46e759f1167f27c7cf49892a42ffa6c1/src/extension.ts#L275
And the user can click the `View logs` button to see the entire error if necessary.

From what I understood this PR was more about which message to show the user in the popup window, and that the AWS errors we choose are known to have sufficient messages so we don't need to give it extra context. If we prepend the Toolkit message I think it will add more noise

Eg: The example given in the original ticket would now look like :
`Access token limit exceeded: Dev Environment usage limit reached for space Space-2`
when 
`Dev Environment usage limit reached for space Space-2` would be sufficient imo.

---

I don't have an issue updating to this proposal though, just wanted to make sure we were on the same page first.

Reply:
> From what I understood this PR was more about which message to show the user in the popup window

It's not so much about which error to show but rather how much we want to show in the notification. Showing everything might be noisy which is why we delegate to the logs, showing only the top-level error message. The top-level error is used because it's the closest to the user flow and has context about the user's actions or goals. 

When there's an error, I want to know _what_ failed and _why_ it failed. Right now we show the "what" but likely not the "why". What you've implemented is basically a heuristic for the "why" part. So, to me, it makes sense to combine the messages rather than show one or the other. They are not mutually exclusive.

>Eg: The example given in the original ticket would now look like :
Access token limit exceeded: Dev Environment usage limit reached for space Space-2
when
Dev Environment usage limit reached for space Space-2 would be sufficient imo.

I believe the error would actually be `Failed to create Dev Environment: Dev Environment usage limit reached for space Space-2` atm which is a lot more relevant than "Access token limit exceeded". But in any case this is a _specific_ scenario while we're writing _generic_ logic. General solutions are not meant to be perfect in every single scenario. They are meant to address the common case while hopefully offering some flexibility for special cases. CodeCatalyst errors may be sufficient to show on their own, but would that be true for other errors with code `ServiceQuotaExceededException`?

Reply:
Thanks for the detail. The 'what' vs 'why' makes sense. Will update!